expression/integration_test.go

```diff
@@ -5997,6 +5997,21 @@ func (s *testIntegrationSerialSuite) TestIssue15315(c *C) {
 	tk.MustQuery("select cast('0-1234' as real)").Check(testkit.Rows("0"))
 }
 
+func (s *testIntegrationSerialSuite) TestIssue23506(c *C) {
+	collate.SetNewCollationEnabledForTest(true)
+	defer collate.SetNewCollationEnabledForTest(false)
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	// FIXME: MySQL will get an error [HY000][3854] Cannot convert string '\x80' from binary to utf8mb4, at this moment,
+	// we can only get an error like this, will try to fix in the future.
+	tk.MustGetErrMsg("select 'a' collate utf8mb4_general_ci = 0x80;", "[expression:1267]Illegal mix of collations (utf8mb4_general_ci,EXPLICIT) and (binary,COERCIBLE) for operation '='")
+	tk.MustExec("drop table if exists t1")
+	tk.MustExec("create table t1(a char(10), primary key (a)) CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;")
+	tk.MustGetErrMsg("select * from t1 where a > 0x80;", "[expression:1267]Illegal mix of collations (utf8mb4_general_ci,IMPLICIT) and (binary,COERCIBLE) for operation '>'")
```

Comment:
There is no error in MySQL, just a warning, do we need to comment on this situation?
```
mysql> select * from t1 where a > 0x80;
Empty set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------------------------+
| Level   | Code | Message                                                |
+---------+------+--------------------------------------------------------+
| Warning | 1366 | Incorrect string value: '\x80' for column 'a' at row 1 |
+---------+------+--------------------------------------------------------+
1 row in set (0.00 sec)
```

Reply:
updated, I should insert some data into the table first, then we will get the error, still, the error is not the same with MySQL